### PR TITLE
rbac: Add instancetype.kubevirt.io:view ClusterRole

### LINKED ISSUE #10437: The `kubevirt:view` `ClusterRole` does not provide permissions for users to access `VirtualMachineCluster{Instancetype,Preference}` cluster scoped resources when bound using `RoleBinding`
/area instancetype
/assign

**What happened**:

$subject, the `kubevirt:view` `ClusterRole` doesn't provide access to cluster scoped resources when bound through a `RoleBinding`:

```
$ oc whoami
kube:admin

$ oc get user/test -o json 
{
    "apiVersion": "user.openshift.io/v1",
    "groups": null,
    "identities": [
        "test:test"
    ],
    "kind": "User",
    "metadata": {
        "creationTimestamp": "2023-09-13T06:32:27Z",
        "name": "test",
        "resourceVersion": "2296746",
        "uid": "82589434-61c0-43f9-b583-4cdf1c07a449"
    }
}
$ oc get rolebinding/view -o json 
{
    "apiVersion": "rbac.authorization.k8s.io/v1",
    "kind": "RoleBinding",
    "metadata": {
        "creationTimestamp": "2023-09-12T00:26:26Z",
        "name": "view",
        "namespace": "default",
        "resourceVersion": "299686",
        "uid": "41047a8a-9d95-42bc-9f59-667ae114ea5c"
    },
    "roleRef": {
        "apiGroup": "rbac.authorization.k8s.io",
        "kind": "ClusterRole",
        "name": "view"
    },
    "subjects": [
        {
            "apiGroup": "rbac.authorization.k8s.io",
            "kind": "User",
            "name": "test"
        }
    ]
}

$ oc get clusterrole/view -o json | jq -r '.rules[] | select(.apiGroups==["instancetype.kubevirt.io"])'
{
  "apiGroups": [
    "instancetype.kubevirt.io"
  ],
  "resources": [
    "virtualmachineinstancetypes",
    "virtualmachineclusterinstancetypes",
    "virtualmachinepreferences",
    "virtualmachineclusterpreferences"
  ],
  "verbs": [
    "get",
    "list",
    "watch"
  ]
}

$ oc whoami
test 

$ oc get virtualmachineclusterinstancetypes
Error from server (Forbidden): virtualmachineclusterinstancetypes.instancetype.kubevirt.io is forbidden: User "test" cannot list resource "virtualmachineclusterinstancetypes" in API group "instancetype.kubevirt.io" at the cluster scope
```

**What you expected to happen**:

A separate `ClusterRole` should be provided by KubeVirt that can be bound with a `ClusterRoleBinding` allowing access to these cluster scoped resources.

**How to reproduce it (as minimally and precisely as possible)**:
As above.

**Additional context**:

**Environment**:
- KubeVirt version (use `virtctl version`): N/A
- Kubernetes version (use `kubectl version`): N/A
- VM or VMI specifications: N/A
- Cloud provider or hardware configuration: N/A
- OS (e.g. from /etc/os-release): N/A
- Kernel (e.g. `uname -a`): N/A
- Install tools: N/A
- Others: N/A

### PR DESCRIPTION
**Pull request by lyarwood**
/area instancetype

**What this PR does / why we need it**:

This ClusterRole provides read only access to the cluster-scoped VirtualMachineCluster{Instancetype,Preference} resources within a cluster.

This is required as the existing aggregated kubevirt:view ClusterRole does not provide access to cluster-scoped resources when bound to a user using a RoleBinding.

As such we need a seperate ClusterRole to provide access to our cluster scoped resources that can be bound with a ClusterRoleBinding without providing access to all of the resources listed in the aggregated view ClusterRole.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #10437

**Special notes for your reviewer**:

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [x] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [x] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [x] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [x] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [x] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:

```release-note
A new `instancetype.kubevirt.io:view` `ClusterRole` has been introduced that can be bound to users via a `ClusterRoleBinding` to provide read only access to the cluster scoped `VirtualMachineCluster{Instancetype,Preference}` resources.
```

### CODE REVIEW

**Comment by lyarwood** [comment]
/hold

I need to confirm this isn't a downstream bug in OKD/OCP given the reference to the downstream User kind.

**Comment by lyarwood** [comment]
/retest-required
/unhold

I was able to reproduce this upstream away from OCP and verified the new `instancetype.kubevirt.io:view` `ClusterRole` worked as expected once applied to my user.

File: tests/testsuite/serviceaccount.go [hunk 1]
```diff
@@ -50,6 +51,9 @@ func createServiceAccounts() {
 	createServiceAccount(ViewServiceAccountName)
 	createRoleBinding(ViewServiceAccountName, "kubevirt.io:view")
 
+	createServiceAccount(ViewInstancetypeServiceAccountName)
+	createClusterRoleBinding(ViewInstancetypeServiceAccountName, "instancetype.kubevirt.io:view")
```
Comment (0xFelix) [inline tests/testsuite/serviceaccount.go:55]:
```suggestion
	createClusterRoleBinding(fmt.Sprintf("%s-%d", ViewInstancetypeServiceAccountName, GinkgoParallelProcess()), "instancetype.kubevirt.io:view")
```

When running tests in parallel you need to give the created `ClusterRoleBindings` individual names, as they are on the cluster level but refer to ServiceAccounts in individual namespaces. This should solve the failing tests. Note: Cleanup needs also to be fixed.

File: tests/testsuite/serviceaccount.go [hunk 2]
```diff
@@ -190,6 +225,11 @@ func cleanupServiceAccount(saName string) {
 		util.PanicOnError(err)
 	}
 
+	err = virtCli.RbacV1().ClusterRoles().Delete(context.Background(), saName, metav1.DeleteOptions{})
```
Comment (0xFelix) [inline tests/testsuite/serviceaccount.go:228]: Make sure to only delete the ClusterRole when all workers are finished.

File: tests/testsuite/serviceaccount.go [hunk 3]
```diff
@@ -199,4 +239,9 @@ func cleanupServiceAccount(saName string) {
 	if !k8serrors.IsNotFound(err) {
 		util.PanicOnError(err)
 	}
+
+	err = virtCli.RbacV1().ClusterRoleBindings().Delete(context.Background(), saName, metav1.DeleteOptions{})
```
Comment (0xFelix) [inline tests/testsuite/serviceaccount.go:243]: Make sure to only delete the ClusterRoleBinding created by this worker.

**Review by 0xFelix — COMMENTED** [review]
There are issues with cluster wide resources in the test code you added. Because CI is running the tests in parallel on multiple workers you need to be careful when creating or deleting cluster wide resources.

File: pkg/virt-operator/resource/generate/rbac/cluster.go [hunk 4]
```diff
@@ -631,3 +632,28 @@ func newViewClusterRole() *rbacv1.ClusterRole {
 		},
 	}
 }
+func newInstancetypeViewClusterRole() *rbacv1.ClusterRole {
```
Comment (xpivarc) [inline pkg/virt-operator/resource/generate/rbac/cluster.go:635]: This is just a role and there is no binding, am I missing anything?

**Comment by jcanocan** [comment]
Great job! :)
/lgtm

**Pull request merged**